### PR TITLE
Fix setter of FunctionSet.parameters and FunctionSet.gradients

### LINKED ISSUE #502: The behavior of Function#parameters= is inconsistent with FunctionSet#parameters=
At aaf2aae14ad683734964590e6ceee9d775d9d7c5, an assertion was introduced into `Function#parameters=`, which requires the argument `values` to be a list-like object (implementing `__len__` method).
However, this assertion seems to be inconsistent with the behavior of `FunctionSet#parameters=`.
`FunctionSet#parameters=` passes the iterator of its argument `params` to its child functions through `Function#parameters=`, and it causes a TypeError exception.

```python
from chainer import FunctionSet
import chainer.functions as F
func_set = FunctionSet(f=F.Linear(10,10))
params = func_set.parameters
func_set.parameters = params # TypeError: object of type 'tuple_iterator' has no len()
```

### PR DESCRIPTION
**Pull request by delta2323**
Fix #502. `FunctionSet.gradients` have same problems as `FunctionSet.parameters`. This PR also fixes `gradients` .

### CODE REVIEW

File: chainer/function_set.py [hunk 1]
```diff
@@ -126,9 +126,12 @@ def parameters(self):
...
     @parameters.setter
     def parameters(self, params):
-        param_iter = iter(params)
+        idx = 0
         for _, func in self._get_sorted_funcs():
-            func.parameters = param_iter
+            param_size = len(func.parameters)
+            func.parameters = params[idx: idx + param_size]
+            idx += param_size
+        assert idx == len(params)
```
Comment (unnonouno) [inline chainer/function_set.py]: Why don't you check sum of length of params before assigning? Such code is exception-safe.

**Comment by unnonouno** [comment]
LGTM except one comment.

File: chainer/function_set.py (hunk 1 above)
Comment (delta2323) [inline chainer/function_set.py]: I agree with you. I will change to check the assertion before assignment.

**Comment by delta2323** [comment]
I fixed the point @unnonouno pointed out.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**